homeassistant/components/switchbot/__init__.py

```diff
@@ -1 +1,97 @@
-"""The switchbot component."""
+"""Support for Switchbot devices."""
+from asyncio import Lock
+
+import switchbot  # pylint: disable=import-error
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import (
+    CONF_RETRY_COUNT,
+    CONF_RETRY_TIMEOUT,
+    CONF_SCAN_TIMEOUT,
+    CONF_TIME_BETWEEN_UPDATE_COMMAND,
+    DATA_COORDINATOR,
+    DATA_UNDO_UPDATE_LISTENER,
+    DEFAULT_RETRY_COUNT,
+    DEFAULT_RETRY_TIMEOUT,
+    DEFAULT_SCAN_TIMEOUT,
+    DEFAULT_TIME_BETWEEN_UPDATE_COMMAND,
+    DOMAIN,
+)
+from .coordinator import SwitchbotDataUpdateCoordinator
+
+PLATFORMS = ["switch"]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Switchbot from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    if not entry.options:
+        options = {
+            CONF_TIME_BETWEEN_UPDATE_COMMAND: DEFAULT_TIME_BETWEEN_UPDATE_COMMAND,
+            CONF_RETRY_COUNT: DEFAULT_RETRY_COUNT,
+            CONF_RETRY_TIMEOUT: DEFAULT_RETRY_TIMEOUT,
+            CONF_SCAN_TIMEOUT: DEFAULT_SCAN_TIMEOUT,
+        }
+
+        hass.config_entries.async_update_entry(entry, options=options)
+
+    # Use same coordinator instance for all entities.
+    # Uses BTLE advertisement data, all Switchbot devices in range is stored here.
+    if hass.data.get(DOMAIN):
+        for item in hass.config_entries.async_entries(domain=DOMAIN):
+            if hass.data[DOMAIN].get(item.entry_id):
+                coordinator = hass.data[DOMAIN][item.entry_id].get(DATA_COORDINATOR)
+                break
```

Comment:
1) Check if coordinator is already loaded by another entity setup.
2) If found, set coordinator to same object.

Data is obtained from BTLE advertisements provided from a scan. Data for all devices in proximity is returned. The API builds a dict with entities and their corresponding setting. (On/Off state and mode for the bots as example)

I'm going to add sensors for Battery etc. in future pull requests.

Reply:
It's not possible to do more than one simultaneous scan per Bluetooth adapter and it's not needed. 

I've tested removing/adding/changing options of entities without any issues or lockups. Can even remove the first loaded entity and others still update just fine.